src/netlink.c

```diff
@@ -0,0 +1,365 @@
+/*-------------------------------------------------------------------------*\
+* Send data through connected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_send(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int flags = luaL_optinteger(L, 3, 0);
+    struct nlmsghdr nlh[NLMSG_SPACE(payload_size)];
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+    timeout_markstart(tm);
+
+    err = socket_send(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, tm);
+    if (err != IO_DONE) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error sending message");
+        return 2;
+    }
+
+    lua_pushinteger(L, sent);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Send data through unconnected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_sendto(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int dstpid = luaL_checkinteger(L, 3);
+    int groups = luaL_optinteger(L, 4, 0);
+    int flags = luaL_optinteger(L, 5, 0);
+    struct nlmsghdr nlh[NLMSG_SPACE(payload_size)];
+    struct sockaddr_nl addr;
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    memset(&addr, 0, sizeof(addr));
+    addr.nl_pid = dstpid;
+    addr.nl_family = AF_NETLINK;
+    addr.nl_groups = groups;
+
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+    timeout_markstart(tm);
+
+    err = socket_sendto(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, 
+            (SA *)&addr, sizeof(addr), tm);
+    if (err != IO_DONE) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error sending message");
+        return 2;
+    }
+
+    lua_pushinteger(L, sent);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    struct nlmsghdr nlh[NLMSG_SPACE(MAX_PAYLOAD)];
+    size_t got;
+    p_timeout tm = &nl->tm;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
```

Comment:
That size is not what the application expects to receive, it's just an arbitrary value, that's sufficiently large to store any message that may come. Much like the 1024 size for DNS messages in the appliance code.

Reply:
Who ever said `MAX_PAYLOAD` is what the application expects to receive? Would you care to enlighten me by point me the documentation or whatever thing from which you took this nonsense about 1024 and DNS?

Reply:
As far as I understand, you did. Here's what I understood from your sentence: Why instead of defining a ceiling value for the message length, thats about to be received, you send that value instead? And about that DNS size, it's not nonsense. If I got anything wrong about that, it was the actual exact size. I'm not sure you remember when me, @VictorNogueiraRio and @LeoLAR found a bug where DNS messages were not being processed, and instead an error was being thrown that they were too large, but @marcelstanley probably does, since he asked us then, if we could reproduce it. It was some sort of sanity check I believe.